cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/NotificationSubscriberService.java

```diff
@@ -0,0 +1,341 @@
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.cConf = cConf;
+    this.messagingService = messagingService;
+    this.topics = cConf.getTrimmedStrings(Constants.AppFabric.SCHEDULER_MESSAGING_TOPICS);
+    this.subscriberThreads = new ArrayList<>();
+    this.readyJobs = new LinkedBlockingQueue<>();
+    this.runFlag = new AtomicBoolean();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    taskExecutorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(Threads.createDaemonThreadFactory("notification-subscriber-task")));
+    taskRunner = new ScheduleTaskRunner(store, lifecycleService, propertiesResolver,
+                                        taskExecutorService, namespaceQueryAdmin, cConf);
+  }
+
+  /**
+   * Temporary workaround when Schedule store is not ready.
+   *
+   * @param scheduleMap Map storing keys constructed from {@link Notification}
+   *                    and corresponding {@link ProgramSchedule} as values
+   */
+  public void setScheduleMap(Map<String, List<ProgramSchedule>> scheduleMap) {
+    this.scheduleMap = scheduleMap;
+  }
+
+  @Override
+  protected void run() {
+    LOG.info("Start running NotificationSubscriberService");
+    if (!isRunning()) {
+      return;
+    }
+
+    for (String topic : topics) {
+      subscriberThreads.add(new NotificationSubscriberThread(topic, null));
+    }
+
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      thread.start();
+    }
+
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      try {
+        thread.join();
+      } catch (InterruptedException e) {
+        LOG.info("Thread {} is being terminated while waiting for it to finish.", thread.getName());
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  @Override
+  protected void triggerShutdown() {
+    LOG.info("Stopping NotificationSubscriberService.");
+    for (NotificationSubscriberThread thread : subscriberThreads) {
+      thread.interrupt();
+    }
+    if (taskExecutorService != null) {
+      taskExecutorService.shutdownNow();
+    }
+    LOG.info("NotificationSubscriberService stopped.");
+  }
+
+  private class NotificationSubscriberThread extends Thread {
+    private final String topic;
+    private final RetryStrategy scheduleStrategy;
+    private final MultiThreadMessagingContext messagingContext;
+    private int emptyFetchCount;
+    private int failureCount;
+    private String messageId;
+
+    NotificationSubscriberThread(String topic, @Nullable String messageId) {
+      super(String.format("NotificationSubscriberThread-%s", topic));
+      this.topic = topic;
+      this.messageId = messageId;
+      // Retry with delay ranging from 0.1s to 30s
+      scheduleStrategy =
+        co.cask.cdap.common.service.RetryStrategies.exponentialDelay(100, 30000, TimeUnit.MILLISECONDS);
+      this.messagingContext = new MultiThreadMessagingContext(messagingService);
```

Comment:
Shouldn't need in here. Should create one in the outer class and provide it as extra tx aware to the `MultiThreadDatasetCache`. Each fetcher thread should just use it. Otherwise the consumption is not transactional.